### PR TITLE
SED-2344 Implement scheduler exclusions

### PR DESCRIPTION
**Pull request by david-stephan**

### CODE REVIEW

File: step-controller/step-controller-server/src/main/java/step/core/scheduler/Executor.java [hunk 1]
```diff
@@ -127,12 +142,45 @@ public boolean schedule(ExecutiontTaskParameters task) {
...
+					baseCalendar = new CronCalendar(baseCalendar, c.getCronExpression());
+				}
+			} catch (ParseException e) { //such exception should already be caught by validate
+				logAndThrow("One of the cron expressions for the task '" + task.getAttribute(AbstractOrganizableObject.NAME) +
+						"' is invalid.", e);
+			}
+		}
+		//In case exclusions were removed we need to add/replace at least the base calendar
+		try {
+			scheduler.addCalendar("exclusionsCalendar_" + task.getId().toHexString(), baseCalendar, true, false);
```
Comment (jeromecomte) [inline step-controller/step-controller-server/src/main/java/step/core/scheduler/Executor.java:159]: Just for my understanding: where do we specify that this calendar is an exclusion calendar? is it implicit in the method modifiedByCalendar?

Comment (jeromecomte) [inline step-controller/step-controller-server/src/main/java/step/core/scheduler/Executor.java:159]: Shouldn't we delete these calendars when a task is deleted?

File: step-controller/step-controller-server/src/main/java/step/core/scheduler/Executor.java [hunk 2]
```diff
@@ -127,12 +142,45 @@ public boolean schedule(ExecutiontTaskParameters task) {
...
+		} catch (SchedulerException e) {//exception throw when adding exclustion calendar
+			logAndThrow("Adding exclusion CRON expressions raised an error for the task '" +
+				task.getAttribute(AbstractOrganizableObject.NAME) + "'.", e);
+		}
+
+		Trigger trigger = TriggerBuilder.newTrigger()
+				.withSchedule(CronScheduleBuilder.cronSchedule(task.getCronExpression()))
+				.modifiedByCalendar("exclusionsCalendar_" + task.getId().toHexString())
```
Comment (jeromecomte) [inline step-controller/step-controller-server/src/main/java/step/core/scheduler/Executor.java:167]: Use a constant for "exclusionsCalendar_" and a function for creating the exclusion calendar id

File: step-controller/step-controller-server/src/main/java/step/core/scheduler/Executor.java [hunk 3]
```diff
@@ -127,12 +142,45 @@ public boolean schedule(ExecutiontTaskParameters task) {
...
+				.build();
 		JobDetail job = buildScheduledJob(task);
-		scheduleJob(trigger, job);
+		try {
+			scheduleJob(trigger, job);
+		} catch (Exception e) {
+			logAndThrow("Unable to schedule task '" + task.getAttribute(AbstractOrganizableObject.NAME) + "'.", e);
```
Comment (jeromecomte) [inline step-controller/step-controller-server/src/main/java/step/core/scheduler/Executor.java:173]: Would create a method for "task.getAttribute(AbstractOrganizableObject.NAME)". This is called at different places in this class

File: step-controller/step-controller-server/src/test/java/step/core/scheduler/ExecutionSchedulerCronTest.java [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+import step.core.controller.InMemoryControllerSettingAccessor;
+import step.core.execution.model.ExecutionParameters;
+import step.core.plugins.PluginManager.Builder.CircularDependencyException;
+
+import java.util.List;
+import java.util.Properties;
+
+import static org.junit.Assert.*;
+
+public class ExecutionSchedulerCronTest {
```
Comment (jeromecomte) [inline step-controller/step-controller-server/src/test/java/step/core/scheduler/ExecutionSchedulerCronTest.java:22]: Nice!

File: step-controller/step-controller-server/src/test/java/step/core/scheduler/ExecutionSchedulerCronTest.java [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+
+		//Test simple CRON without exclutions
+		final ExecutiontTaskParameters executiontTaskParameters = new ExecutiontTaskParameters();
+		executiontTaskParameters.addAttribute(AbstractOrganizableObject.NAME, "task1");
+		executiontTaskParameters.setExecutionsParameters(new ExecutionParameters());
+		executiontTaskParameters.setCronExpression("0 0/1 * * * FRIDAY");
+		assertThrows("CronExpression '0 0/1 * * * FRIDAY' is invalid. Details: Support for specifying both a day-of-week AND a day-of-month parameter is not implemented.",
+				RuntimeException.class, () -> executionScheduler.addExecutionTask(executiontTaskParameters));
+
+		//Test with same CRON for scheduling and excluding
```
Comment (jeromecomte) [inline step-controller/step-controller-server/src/test/java/step/core/scheduler/ExecutionSchedulerCronTest.java:123]: Commented code

**Review by jeromecomte — COMMENTED** [review]
The PR looks good. Waiting for a few clarifications to approve.
One general question: now that we're using cron expressions to define exclusions, is the plan to expose these expression to the end user too? if yes, it might be helpful to have a kind of wizard to help the user create these expressions

File: step-controller/step-controller-server/src/main/java/step/core/scheduler/Executor.java [hunk 6]
```diff
@@ -114,7 +122,14 @@ public void deleteSchedule(ExecutiontTaskParameters task) {
 	}
 	
 	public void validate(ExecutiontTaskParameters task) {
-		CronScheduleBuilder.cronSchedule(task.getCronExpression());
+		try {
+			CronScheduleBuilder.cronSchedule(task.getCronExpression());
+			if (task.getCronExclusions() != null) {
+				task.getCronExclusions().forEach(c-> CronScheduleBuilder.cronSchedule(c.getCronExpression()));
```
[the remark this answers is not in the document]
Comment (david-stephan) [reply step-controller/step-controller-server/src/main/java/step/core/scheduler/Executor.java:128]: @jeromecomte yes these are the same expressions and the same way to validate them, range are supported by CRON expression

File: step-controller/step-controller-server/src/main/java/step/core/scheduler/Executor.java [hunk 7]
```diff
@@ -114,7 +122,14 @@ public void deleteSchedule(ExecutiontTaskParameters task) {
...
 	
 	public void validate(ExecutiontTaskParameters task) {
-		CronScheduleBuilder.cronSchedule(task.getCronExpression());
+		try {
+			CronScheduleBuilder.cronSchedule(task.getCronExpression());
+			if (task.getCronExclusions() != null) {
+				task.getCronExclusions().forEach(c-> CronScheduleBuilder.cronSchedule(c.getCronExpression()));
+			}
+		} catch (RuntimeException e) {
+			logAndThrow(e.getMessage(), e);
```
[the remark this answers is not in the document]
Comment (david-stephan) [reply step-controller/step-controller-server/src/main/java/step/core/scheduler/Executor.java:131]: @jeromecomte the expression handling has been improved as part of this CR and now contains the CRON expression as illustrated in the new Junit test

File: step-controller/step-controller-server/src/main/java/step/core/scheduler/Executor.java (hunk 1 above)
Comment (david-stephan) [reply step-controller/step-controller-server/src/main/java/step/core/scheduler/Executor.java:159]: @jeromecomte in the Trigger builder ".modifiedByCalendar"

Comment (david-stephan) [reply step-controller/step-controller-server/src/main/java/step/core/scheduler/Executor.java]: @jeromecomte yes, added in new commit

File: step-controller/step-controller-server/src/main/java/step/core/scheduler/Executor.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (david-stephan) [reply step-controller/step-controller-server/src/main/java/step/core/scheduler/Executor.java]: @jeromecomte implemented in new commit

File: step-controller/step-controller-server/src/main/java/step/core/scheduler/Executor.java (hunk 3 above)
Comment (david-stephan) [reply step-controller/step-controller-server/src/main/java/step/core/scheduler/Executor.java]: refactored with variable and static import. Refactoring with a method should be done directly on AbstractOrganizableObject.

File: step-controller/step-controller-server/src/test/java/step/core/scheduler/ExecutionSchedulerCronTest.java (hunk 5 above)
Comment (david-stephan) [reply step-controller/step-controller-server/src/test/java/step/core/scheduler/ExecutionSchedulerCronTest.java:127]: cleaned up

**Comment by david-stephan** [comment]
@jeromecomte individual points answered/addressed. A picker is already planned, the user will still have the possibility to enter them manually/directly as for the "scheduling" CRON expressions.

**Review by jeromecomte — APPROVED** [review]

**Pull request closed without merging**